### PR TITLE
[dashboard] usage view allows for flexible time range

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description
This change allows to set custom date ranges on the usage view.
Screenshot 2022-11-08 at 20 49 27

## Related Issue(s)

See https://github.com/gitpod-io/gitpod/issues/14463, https://github.com/gitpod-io/gitpod/issues/13288

## How to test

- create some usage
- go to https://se-usage-view.preview.gitpod-dev.com/usage
- Also try https://se-usage-view.preview.gitpod-dev.com/usage#2022-01-08:2022-01-12

## Release Notes

```release-note
Usage view allows for arbitrary date ranges
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

### CODE REVIEW

**Comment by roboquat** [comment]
@svenefftinge: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/dashboard/src/components/UsageView.tsx [hunk 1]
```diff
@@ -36,22 +40,37 @@ function UsageView({ attributionId }: UsageViewProps) {
...
+    const [startDate, setStartDate] = useState(timestampStartOfCurrentMonth);
+    const [startDateError, setStartDateError] = useState("");
+    const [endDate, setEndDate] = useState(Date.now());
     const [totalCreditsUsed, setTotalCreditsUsed] = useState<number>(0);
     const [isLoading, setIsLoading] = useState<boolean>(true);
     const [supportedClasses, setSupportedClasses] = useState<SupportedWorkspaceClass[]>([]);
 
+    const location = useLocation();
     useEffect(() => {
+        const match = /#(\d{4}-\d{2}-\d{2}):(\d{4}-\d{2}-\d{2})/.exec(location.hash);
```
Comment (easyCZ) [inline components/dashboard/src/components/UsageView.tsx:52]: What's the reason to use a hash and not standard query params such as `from` and `to`?

The Regex looks to only match days, without timestamps. Do we want to allow for time also?

We already use `dayjs` in the project, perhaps using https://day.js.org/docs/en/parse/string-format would be cleaner here?

File: components/dashboard/src/components/UsageView.tsx [hunk 2]
```diff
@@ -36,22 +40,37 @@ function UsageView({ attributionId }: UsageViewProps) {
...
+            }
+        }
         (async () => {
             const classes = await getGitpodService().server.getSupportedWorkspaceClasses();
             setSupportedClasses(classes);
         })();
-    }, []);
-
+    }, [location]);
+    // edit
```
Comment (easyCZ) [inline components/dashboard/src/components/UsageView.tsx:66]: remove?

File: components/dashboard/src/components/UsageView.tsx [hunk 3]
```diff
@@ -36,22 +40,37 @@ function UsageView({ attributionId }: UsageViewProps) {
     const today = new Date();
     const startOfCurrentMonth = new Date(today.getFullYear(), today.getMonth(), 1);
```
Comment (easyCZ) [inline components/dashboard/src/components/UsageView.tsx:41]: [drive-by] Could be simplified to `dayjs().startOf('month')`, we already do this [here](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/src/components/UsageBasedBillingConfig.tsx#L48)

**Comment by easyCZ** [comment]
Nice improvement!

**Review by geropl — APPROVED** [review]
Tested and works nicely. :sparkling_heart: 

/hold in case you want to handle [this](https://github.com/gitpod-io/gitpod/pull/14515/files#r1017052770) here, now.

File: components/dashboard/src/components/UsageView.tsx (hunk 1 above)
Comment (svenefftinge) [reply components/dashboard/src/components/UsageView.tsx:52]: I intentionally opted for has instead of URL params, because this is entirely clientside.

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/components/UsageView.tsx [hunk 4]
```diff
@@ -160,13 +179,61 @@ function UsageView({ attributionId }: UsageViewProps) {
...
-                ).toLocaleDateString()} (updated every 15 minutes).`}
+                title={
+                    <div className="flex items-baseline">
+                        <h1 className="tracking-wide">{headerTitle}</h1>
+                        <h2 className="ml-3">(updated every 15 minutes).</h2>
+                    </div>
+                }
+                subtitle={
+                    <h2 className="tracking-wide flex mt-3">
+                        <h2 className="mr-2">Period: </h2>
```
Comment (gtsiolis) [inline components/dashboard/src/components/UsageView.tsx:212]: **thought**: The idea of having the period selection on the top of the left sidebar is to update all usage entries on the main content as well as all aggregated data on the left sidebar based on the period selection. We could use the heading subtitle for hijacking this action and filtering other views like workpaces per team, etc.

| Idea of the top-left control for period selection | 
|-|
| Frame 1135 |

Comment (gtsiolis) [inline components/dashboard/src/components/UsageView.tsx:212]: **question**: While the goal long term is to replace the primitive navigation between months with more flexible options like Last 7 days, 30 days, etc with a flexible dropdown (see https://github.com/gitpod-io/gitpod/issues/11685#issuecomment-1197939139) to keep the scope small here, what do you think of moving the period elements below inside the left sidebar? Alternatively, we could keep the custom filtering on the heading subtitle which feels nice.

| Range picker on the sidebar | Range picker on the heading |
|-|-|
| Screenshot 2022-11-09 at 9 22 57 PM | Screenshot 2022-11-09 at 9 27 59 PM |

File: components/dashboard/src/components/UsageView.tsx [hunk 5]
```diff
@@ -160,13 +179,61 @@ function UsageView({ attributionId }: UsageViewProps) {
...
+                            customInput={<FromLink />}
+                            dateFormat={"dd/MM/yyyy"}
+                        />
+                        <div className="mx-2">-</div>
+                        <DatePicker
+                            selected={new Date(endDate)}
+                            onChange={(date) => date && setEndDate(date.getTime())}
+                            selectsEnd
+                            startDate={new Date(startDate)}
+                            endDate={new Date(endDate)}
```
Comment (gtsiolis) [inline components/dashboard/src/components/UsageView.tsx:230]: **suggestion**: Could we show the last day of a month for each month selection? Currently, for example, when selecting October 2022, the range changes to `01/10/2022 - 01/11/2022`, and would be more accurate if this would change to `01/10/2022 - 31/10/2022`.

File: components/dashboard/src/components/UsageView.tsx [hunk 6]
```diff
@@ -160,13 +179,61 @@ function UsageView({ attributionId }: UsageViewProps) {
...
+                            maxDate={new Date(endDate)}
+                            customInput={<FromLink />}
+                            dateFormat={"dd/MM/yyyy"}
+                        />
+                        <div className="mx-2">-</div>
+                        <DatePicker
+                            selected={new Date(endDate)}
+                            onChange={(date) => date && setEndDate(date.getTime())}
+                            selectsEnd
+                            startDate={new Date(startDate)}
```
Comment (gtsiolis) [inline components/dashboard/src/components/UsageView.tsx:229]: **suggestion**: Since we are here, not sure if this is can be easily done with the selected library but what do you think of going with a friendlier rendering like `Nov 1, 2022 - Nov 9, 2022` instead of `01/11/2022 - 09/11/2022`?

File: components/dashboard/src/components/UsageView.tsx [hunk 7]
```diff
@@ -160,13 +179,61 @@ function UsageView({ attributionId }: UsageViewProps) {
...
         <>
             <Header
-                title={headerTitle}
-                subtitle={`${new Date(startDateOfBillMonth).toLocaleDateString()} - ${new Date(
-                    endDateOfBillMonth,
-                ).toLocaleDateString()} (updated every 15 minutes).`}
+                title={
+                    <div className="flex items-baseline">
+                        <h1 className="tracking-wide">{headerTitle}</h1>
+                        <h2 className="ml-3">(updated every 15 minutes).</h2>
```
Comment (gtsiolis) [inline components/dashboard/src/components/UsageView.tsx:207]: **suggestion**: This one hurt a bit! 😁 What do you think of moving this information to the sidebar?

Screenshot 2022-11-09 at 9 37 00 PM

File: components/dashboard/src/components/UsageView.tsx [hunk 8]
```diff
@@ -160,13 +179,61 @@ function UsageView({ attributionId }: UsageViewProps) {
...
+                    </div>
+                }
+                subtitle={
+                    <h2 className="tracking-wide flex mt-3">
+                        <h2 className="mr-2">Period: </h2>
+                        {startDateError}
```
Comment (gtsiolis) [inline components/dashboard/src/components/UsageView.tsx:213]: **issue**: Probably not the same error as in the screenshot, and maybe out of the scope of the changes, but we should probably impose a limit on the the max date range gap like 180 days or something, right? A good MVC could be to surface a danger alert with the error message instead of breaking the page.

Screenshot 2022-11-09 at 9 39 34 PM

File: components/dashboard/package.json [hunk 9]
```diff
@@ -9,13 +9,15 @@
     "@gitpod/public-api": "0.1.5",
     "@stripe/react-stripe-js": "^1.7.2",
     "@stripe/stripe-js": "^1.29.0",
+    "@types/react-datepicker": "^4.8.0",
```
Comment (gtsiolis) [inline components/dashboard/package.json:12]: **issue**: Besides all the rendering and interaction issues of this library, I've heard multiple accessibility issues around it, see [relevant blog post](https://adrianroselli.com/2019/07/maybe-you-dont-need-a-date-picker.html#:~:text=Unnecessary%20detail%20about%20one%20library%E2%80%99s%20date%20picker%E2%80%A6). :x:

Comment (gtsiolis) [inline components/dashboard/package.json:12]: **suggestion**: I assume the library we chose helps us bootstrap this, but could we use a better alternative, like the **[date picker component](https://github.com/duetds/date-picker)** from the duet design system, which is also open source? Alternatively, we could go with our current option and try to improve the style and accessibility issues around it.

File: components/dashboard/src/components/react-datepicker.css [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    @apply dark:bg-gray-300 dark:text-gray-800
+}
+
+.react-datepicker div.react-datepicker__day--selecting-range-end {
+    @apply dark:bg-gray-300 dark:text-gray-800
+}
+
+.react-datepicker button {
+    @apply dark:bg-gray-800 dark:text-gray-400
+}
```
Comment (gtsiolis) [inline components/dashboard/src/components/react-datepicker.css:37]: **praise**: Thanks for caring about rending of this in dark theme. 🌔

File: components/dashboard/src/components/UsageView.tsx (hunk 4 above)
Comment (gtsiolis) [inline components/dashboard/src/components/UsageView.tsx:212]: **thought**: Ideally, the custom period selection should be as close as possible with the fixed period selections.

**Review by gtsiolis — COMMENTED** [review]
Thanks @svenefftinge for adding this and hanging here while I was collecting some all around  thoughts around the _hot_ 🔥  topic of a date picker component.

Left some first-round comments below to keep the conversations going.

File: components/dashboard/src/components/UsageView.tsx (hunk 1 above)
Comment (svenefftinge) [reply components/dashboard/src/components/UsageView.tsx:52]: > The Regex looks to only match days, without timestamps. Do we want to allow for time also?

Yes, it does only match days. Why would you use time? We don't display the times in the ranges.

Comment (easyCZ) [reply components/dashboard/src/components/UsageView.tsx:52]: It would replace the regex also, with the date parse string. Basically, we're writing the regex ourselves when it's already been done for us (and tested). Up to you.

Comment (easyCZ) [reply components/dashboard/src/components/UsageView.tsx:52]: > Yes, it does only match days. Why would you use time? We don't display the times in the ranges.

I would encode time as well, for completeness. Yes, it would require a time selection UI, or you could default to midninght on both, unless the URL params say otherwise.

Comment (svenefftinge) [reply components/dashboard/src/components/UsageView.tsx:52]: > It would replace the regex also, with the date parse string. Basically, we're writing the regex ourselves when it's already been done for us (and tested). Up to you.

It would not replace the regex, because the regex includes the delimiter between `<from>:<to>`.

Comment (svenefftinge) [reply components/dashboard/src/components/UsageView.tsx:52]: I find it more readable to be strict, but I'm happy to just match for the delimiter and pass what every is on both sides into `Date.parse` and then validate the resulting dates.

Comment (svenefftinge) [reply components/dashboard/src/components/UsageView.tsx:52]: Maybe even render parsing and validation errors. Although ultimately the protocol is intended to be used by ourselves.

Comment (easyCZ) [reply components/dashboard/src/components/UsageView.tsx:52]: What do you mean `strict`? 

It would be used as `dayjs(queryParam, 'YYYY-MM-DD')` - this is exactly as strict as before, but actually communicates which format it's using by using standard date parse arguments.

File: components/dashboard/package.json (hunk 9 above)
Comment (svenefftinge) [reply components/dashboard/package.json:12]: I tried that lib a little but struggled to make it work with react and typescript.

Comment (svenefftinge) [reply components/dashboard/package.json:12]: The decision for `react-datepicker` is easily revertable as we really only use the element. So if at some point in future we believe it's worth investing in an alternative we can easily replace it.

Comment (gtsiolis) [reply components/dashboard/package.json:12]: DEAL—Let's try to bring the current library to offer a bare minimum good UX. I can try to make some visual changes in a follow-up PR if needed. 🤝

Comment (gtsiolis) [reply components/dashboard/package.json:12]: Thanks for giving this a try! 🔥 😭 

I'd love us to move away from the existing one, but like you [mentioned above](https://github.com/gitpod-io/gitpod/pull/14515#discussion_r1019110025), let's wait until it's worth investing time and effort to design one for our needs. 🤝 

In the meantime, from https://github.com/gitpod-io/gitpod/pull/14515#discussion_r1019119043:
> Let's try to bring the current library to offer a bare minimum good UX. I can try to make some visual changes in a follow-up PR if needed.

File: components/dashboard/src/components/UsageView.tsx (hunk 4 above)
Comment (svenefftinge) [reply components/dashboard/src/components/UsageView.tsx:212]: If you let me pick, I'd go with the "Selection on the heading" version. Looks slickest to me.
Would be great if you could take a general look at how we want to do the range selection out of the scope of this PR.

File: components/dashboard/src/components/UsageView.tsx [hunk 11]
```diff
@@ -160,13 +179,61 @@ function UsageView({ attributionId }: UsageViewProps) {
...
     return (
         <>
             <Header
-                title={headerTitle}
-                subtitle={`${new Date(startDateOfBillMonth).toLocaleDateString()} - ${new Date(
-                    endDateOfBillMonth,
-                ).toLocaleDateString()} (updated every 15 minutes).`}
+                title={
+                    <div className="flex items-baseline">
+                        <h1 className="tracking-wide">{headerTitle}</h1>
```
[the remark this answers is not in the document]
Comment (gtsiolis) [reply components/dashboard/src/components/UsageView.tsx]: Resolving. Thanks for making the change. ☺️

File: components/dashboard/src/components/UsageView.tsx (hunk 7 above)
Comment (gtsiolis) [reply components/dashboard/src/components/UsageView.tsx]: **question**: We dropped this information, but I can open a follow up PR to add this back. Would you say it's not as useful to include in the page? If you think it's ok to include, here's the SVG. Using one gray color could suffice for the first iteration on light and dark theme.

```svg
<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24">
  <path fill="#0F172A" d="M12.75 6c0-.41421-.3358-.75-.75-.75s-.75.33579-.75.75h1.5ZM12 12h-.75c0 .4142.3358.75.75.75V12Zm4.5.75c.4142 0 .75-.3358.75-.75s-.3358-.75-.75-.75v1.5Zm3.75-.75c0 4.5563-3.6937 8.25-8.25 8.25v1.5c5.3848 0 9.75-4.3652 9.75-9.75h-1.5ZM12 20.25c-4.55635 0-8.25-3.6937-8.25-8.25h-1.5c0 5.3848 4.36522 9.75 9.75 9.75v-1.5ZM3.75 12c0-4.55635 3.69365-8.25 8.25-8.25v-1.5c-5.38478 0-9.75 4.36522-9.75 9.75h1.5ZM12 3.75c4.5563 0 8.25 3.69365 8.25 8.25h1.5c0-5.38478-4.3652-9.75-9.75-9.75v1.5ZM11.25 6v6h1.5V6h-1.5Zm5.25 5.25H12v1.5h4.5v-1.5Z"/>
</svg>
```

File: components/dashboard/src/components/UsageView.tsx [hunk 12]
```diff
@@ -160,13 +175,67 @@ function UsageView({ attributionId }: UsageViewProps) {
 
     const headerTitle = attributionId.kind === "team" ? "Team Usage" : "Personal Usage";
 
+    const DateDisplay = forwardRef((arg: any, ref: any) => (
+        <div
+            className="px-2 p-1 text-gray-500 bg-gray-100 dark:text-gray-400 dark:bg-gray-800 rounded-md cursor-pointer flex items-center"
```
Comment (gtsiolis) [inline components/dashboard/src/components/UsageView.tsx:180]: **nitpick**: Minor hover state fix.


```suggestion
            className="px-2 p-1 text-gray-500 bg-gray-50 dark:text-gray-400 dark:bg-gray-800 rounded-md cursor-pointer flex items-center hover:bg-gray-100 dark:hover:bg-gray-700"
```

File: components/dashboard/src/components/UsageView.tsx [hunk 13]
```diff
@@ -160,13 +175,67 @@ function UsageView({ attributionId }: UsageViewProps) {
...
-                    endDateOfBillMonth,
-                ).toLocaleDateString()} (updated every 15 minutes).`}
+                title={
+                    <div className="flex items-baseline">
+                        <h1 className="tracking-tight">{headerTitle}</h1>
+                    </div>
+                }
+                subtitle={
+                    <div className="tracking-wide flex mt-3 items-center">
+                        <h2 className="mr-2">Showing usage from </h2>
```
Comment (gtsiolis) [inline components/dashboard/src/components/UsageView.tsx:215]: **nitpick**: Minor spacing fix.

```suggestion
                        <h2 className="mr-1">Showing usage from </h2>
```

File: components/dashboard/src/components/UsageView.tsx [hunk 14]
```diff
@@ -160,13 +175,67 @@ function UsageView({ attributionId }: UsageViewProps) {
...
+                            selected={startDate.toDate()}
+                            onChange={(date) => date && setStartDate(dayjs(date))}
+                            selectsStart
+                            startDate={startDate.toDate()}
+                            endDate={endDate.toDate()}
+                            maxDate={endDate.toDate()}
+                            customInput={<DateDisplay />}
+                            dateFormat={"MMM d, yyyy"}
+                        />
+                        <h2 className="mx-2">to</h2>
```
Comment (gtsiolis) [inline components/dashboard/src/components/UsageView.tsx:226]: **nitpick**: Minor spacing fix.

```suggestion
                        <h2 className="mx-1">to</h2>
```

File: components/dashboard/src/components/UsageView.tsx (hunk 12 above)
Comment (gtsiolis) [inline components/dashboard/src/components/UsageView.tsx:180]: **nitpick**: Minor line height fix.

```suggestion
            className="px-2 py-0.5 text-gray-500 bg-gray-100 dark:text-gray-400 dark:bg-gray-800 rounded-md cursor-pointer flex items-center"
```

File: components/dashboard/src/components/UsageView.tsx [hunk 15]
```diff
@@ -160,13 +175,67 @@ function UsageView({ attributionId }: UsageViewProps) {
...
-                subtitle={`${new Date(startDateOfBillMonth).toLocaleDateString()} - ${new Date(
-                    endDateOfBillMonth,
-                ).toLocaleDateString()} (updated every 15 minutes).`}
+                title={
+                    <div className="flex items-baseline">
+                        <h1 className="tracking-tight">{headerTitle}</h1>
+                    </div>
+                }
+                subtitle={
+                    <div className="tracking-wide flex mt-3 items-center">
```
Comment (gtsiolis) [inline components/dashboard/src/components/UsageView.tsx:214]: **nitpick**: Minor spacing issue, breaking line height.

```suggestion
                    <div className="tracking-wide flex items-center">
```

Comment (gtsiolis) [inline components/dashboard/src/components/UsageView.tsx:214]: **issue(non-blocking)**: This structure breaks slightly the heigh of the header but let's fix this in https://github.com/gitpod-io/gitpod/issues/6764 unless you'd like to move this element in the `div` above, see subtitle and bottom border.

Frame 1136

File: components/dashboard/src/components/UsageView.tsx [hunk 16]
```diff
@@ -160,13 +175,67 @@ function UsageView({ attributionId }: UsageViewProps) {
...
+                        <DatePicker
+                            selected={endDate.toDate()}
+                            onChange={(date) => date && setEndDate(dayjs(date))}
+                            selectsEnd
+                            startDate={startDate.toDate()}
+                            endDate={endDate.toDate()}
+                            minDate={startDate.toDate()}
+                            customInput={<DateDisplay />}
+                            dateFormat={"MMM d, yyyy"}
+                        />
```
Comment (gtsiolis) [inline components/dashboard/src/components/UsageView.tsx:236]: **question**: Could we easily drop the triangle on calendar expand? Feels like the only remaining design issue.

| Light theme | Dark theme |
|-|-|
| triangle-light | triangle-dark |

**Review by gtsiolis — APPROVED** [review]
Alright, almost there! Thanks for addressing all design-related feedback, @svenefftinge!

Left one more round of comments.

Other than that, looks great!

File: components/dashboard/src/components/react-datepicker.css [hunk 17]
```diff
@@ -0,0 +1,37 @@
+/**
+ * Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+.react-datepicker-wrapper {
```
Comment (gtsiolis) [inline components/dashboard/src/components/react-datepicker.css:7]: **suggestion**: Last comment, promise! If it's not too much to add, `rounded-xl` on the calendar element!

| BEFORE | AFTER |
|-|-|
| calendar-before | calendar-after |

File: components/dashboard/src/components/UsageView.tsx (hunk 15 above)
Comment (svenefftinge) [reply components/dashboard/src/components/UsageView.tsx:214]: Removing this looks bad. Are you sure? If so please fix in a follow up.
